external_search.py

```diff
@@ -704,74 +673,222 @@ def _collections():
             _collections
         )
 
-class ExternalSearchIndexVersions(object):
 
-    VERSIONS = ['v4']
+class MappingDocument(object):
+    """This class knows a lot about how the 'properties' section of an
+    Elasticsearch mapping document (or one of its subdocuments) is
+    created.
+    """
 
-    @classmethod
-    def latest(cls):
-        version_re = re.compile('v(\d+)')
-        versions = [int(re.match(version_re, v).groups()[0]) for v in cls.VERSIONS]
-        latest = sorted(versions)[-1]
-        return 'v%d' % latest
+    def __init__(self):
+        self.properties = {}
+        self.subdocuments = {}
 
-    @classmethod
-    def latest_body(cls):
-        version_method = cls.latest() + '_body'
-        return getattr(cls, version_method)()
+    def add_property(self, name, type, **description):
+        """Add a field to the list of properties.
+
+        :param name: Name of the field as found in search documents.
+        :param type: Type of the field. This may be a custom type,
+            so long as a hook method is defined for that type.
+        :param description: Description of the field.
+        """
+        # TODO: For some fields we could set index: False here, which
+        # would presumably lead to a smaller index and faster
+        # updates. However, it might hurt performance of
+        # searches. When this code is more mature we can do a
+        # side-by-side comparison.
+
+        defaults = dict(index=True, store=False)
+        description['type'] = type
+        for default_name, default_value in defaults.items():
+            if default_name not in description:
+                description[default_name] = default_value
+
+        hook_method = getattr(self, type + "_property_hook", None)
```

Comment:
This `_property_hook` thing is the core innovation of this branch. It allows us to define our own field types so that an actual document or subdocument definition can be represented as a simple dictionary. All the complex behavior goes into the `_property_hook` implementation for a custom type.